draft-ietf-dnsop-svcb-https.md

```diff
@@ -495,59 +495,55 @@ is the effective TargetName:
+2. Issue a SVCB query for $QNAME.
 
-3. If an AliasMode SVCB record is returned for $SVCB_QNAME,
-   clients MUST set $ADDR_QNAME and $SVCB_QNAME to its TargetName (without
+3. If an AliasMode SVCB record is returned, set $QNAME to its TargetName (without
    additional prefixes) and loop back to step 2,
    subject to chain length limits and loop detection heuristics (see
    {{client-failures}}).
 
-4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned
-   for $SVCB_QNAME, clients SHOULD select the highest-priority compatible record.
-   This record's TargetName and SvcParams represent the preferred endpoint.  If
-   connection to this endpoint fails, the client SHOULD try to connect using
-   values from the next-highest-priority compatible record, etc. If all attempts
-   fail, clients SHOULD go to step 5 (except as noted in {{ech-client-behavior}}).
+4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned,
+   these represent the alternative endpoints.
 
-5. At this point there are no usable ServiceMode records, because
-   - there were no SVCB records found for $SVCB_QNAME, OR
-   - all records found were incompatible with this client, OR
-   - all connection attempts using ServiceMode records failed.
-
-   Accordingly, clients SHALL connect to the
-   endpoint consisting of $ADDR_QNAME, the authority endpoint's
-   port number, and no SvcParams.
-
-If all of the above connection attempts fail, clients MAY connect to the
-authority endpoint (except as noted in {{client-failures}} and
-{{ech-client-behavior}}).
+5. Otherwise, SVCB resolution has failed, and the list of known endpoints is
+   empty.
 
 This procedure does not rely on any recursive or authoritative DNS server to
 comply with this specification or have any awareness of SVCB.
 
-When selecting between AAAA and A records to use, clients may use an approach
-such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.
+Once SVCB resolution has concluded, the client proceeds with connection
+establishment.  Clients SHOULD try higher-priority alternatives first, with
+fallback to lower-priority alternatives.  Clients issue AAAA and/or A
+queries for the selected TargetName, and MAY choose between them using an
+approach such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.
+
+A client is called "SVCB-optional" if it can connect without the use of
+ServiceMode records, and "SVCB-reliant" otherwise.  Clients for pre-existing
+protocols SHALL implement SVCB-optional behavior (except as noted in
+{{client-failures}} and {{ech-client-behavior}}).
+
+SVCB-optional clients SHOULD issue in parallel any other DNS queries that might
+be needed for connection establishment.  SVCB-optional clients SHALL append an
+alternative endpoint consisting of the final value of $QNAME, the authority
+endpoint's port number, and no SvcParams, to the list of alternative endpoints, which is 
+attempted before falling back to non-SVCB connection modes.
```

Comment:
The case somewhat ambiguous here is the scenario where an AliasMode record points to a name with only A/AAAA records.  It may help to call this out explicitly.  Which may be clarifying that each time $QNAME changes, SVCB-optional clients SHOULD issue AAAA/A queries.  We may also want some text such as:  "Note that SVCB-optional clients should be prepared to handle the case where the TargetName of an AliasMode SVCB RR has only AAAA/A records and no ServiceMode record."

Reply:
OK, I've added a clarification for that case.